### PR TITLE
Create mini-mailgun

### PR DESCRIPTION
**Pull request by jimbobhickville**
If you want me to flesh out anything that is incomplete, let me know, but I need to get up what I have because it's being too much of a distraction.

Notes:
* Runs in docker via docker-compose
* No unit tests (yet)
* Doesn't actual send the mail (no smtp server, so it just fails every time and requeues until retries are exhausted)
* No support for attachments (unless sent as part of the body)

### CODE REVIEW

File: mailgun/models.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        """
+        msg_dict = self.as_dict(include_body=True)
+        mime_msg = MIMEText(msg_dict['body'])
+        mime_msg['Subject'] = msg_dict['subject']
+        if msg_dict['to']:
+            mime_msg['To'] = ','.join(msg_dict['to'])
+        if msg_dict['cc']:
+            mime_msg['Cc'] = ','.join(msg_dict['cc'])
+        if msg_dict['bcc']:
+            mime_msg['Bcc'] = ','.join(msg_dict['bcc'])
```
Comment (anton-efimenko) [inline mailgun/models.py]: That's not really important but is there any reason you insert `Bcc` header to the MIME? I thought it would defeat the purpose of it.

File: mailgun/models.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        This makes the object serializable by, say, json.dumps()
+
+        :param include_body: flag indicating whether to include larger
+                             message bodies that were put in the BLOB store
+        :type include_body: bool
+        :returns: dict
+        """
+        columns = [prop.key
+                   for prop in inspect(self.__class__).iterate_properties
+                   if isinstance(prop, ColumnProperty)]
```
Comment (anton-efimenko) [inline mailgun/models.py]: I wouldn't say I'm a big fan of it as it makes the implementation implicit but I can see the idea.

File: mailgun/models.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+            record_attempt_data.update(state=const.MessageState.PROCESSING)
+
+        self.update(**record_attempt_data)
+
+        try:
+            email_txt = self.as_email()
+            LOG.info("Sending email:\n%s", email_txt)
+            smtp = smtplib.SMTP(CONF.processor.smtp_host)
+            smtp.set_debuglevel(1)
+            smtp.sendmail(self.sender, self.to, email_txt)
```
Comment (anton-efimenko) [inline mailgun/models.py]: Am I getting it wrong or `cc` and `bcc` wouldn't get a message?

File: mailgun/models.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        """Retrieve a list of retryable messages.
+
+        A 'retryable' message is one that is currently in the REJECTED state
+        and whose next_attempt time has come.
+
+        :returns: list of `~mailgun.models.Message` objects
+        """
+        query = db_session.get_session().query(cls)
+        query = query.filter(Messages.state == const.MessageState.REJECTED)
+        return query.filter(Messages.next_attempt <= datetime.utcnow()).all()
```
Comment (anton-efimenko) [inline mailgun/models.py]: Jugging by the description  `whose next_attempt time has come` it should say

```python
Messages.next_attempt >= datetime.utcnow()
```

**Comment by anton-efimenko** [comment]
Would you change anything if you wanted to spin up 2 processors connecting to the same crate instance?

File: mailgun/api/messages.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            to=msg_json.get('to', []),
+            cc=msg_json.get('cc', []),
+            bcc=msg_json.get('bcc', []),
+        )
+        msg_obj.queue()
+
+        # debated HTTP_CREATED, but I think HTTP_ACCEPTED is a stronger
+        # indication that it's still being processed
+        resp.status = falcon.HTTP_ACCEPTED
+        resp.body = {'resource': msg_obj.as_dict(include_body=False)}
```
Comment (russjones) [inline mailgun/api/messages.py]: Why return the entire message body here?

File: mailgun/models.py (hunk 1 above)
Comment (jimbobhickville) [inline mailgun/models.py]: Yeah, you're right.  I kind of hurried through this part and didn't think about it that much.  I imagine smtplib has a way to do this correctly, and I just glossed over it.  I'll fix up the sending and add an smtp server to verify that it's doing the right thing.

File: mailgun/models.py (hunk 4 above)
Comment (jimbobhickville) [inline mailgun/models.py]: That would get anything that was supposed to requeue in the future.  You want it to get anything with a next_attempt in the past but that hasn't changed state from REJECTED yet.  This lets it automatically catch up if it has an outage, as well.

File: mailgun/api/messages.py (hunk 5 above)
Comment (jimbobhickville) [inline mailgun/api/messages.py]: I suppose we could return the body, but the main goal was that the response would have the generated ```message_id``` so they could use that to poll for status changes, if desired.  I figured without the body would be a faster response since they wouldn't have to re-download the whole message if it was large.

**Comment by jimbobhickville** [comment]
So 2 processors *should* work as-is because Kafka will auto-balance consumers with the same ```group_id``` so they don't get the same messages, but if you wanted to be extra careful, I would add some sort of mutex on the crate records to avoid race conditions with multiple processors updating the same record.

File: mailgun/models.py (hunk 3 above)
Comment (jimbobhickville) [inline mailgun/models.py]: You might be right, the docs for ```smtplib``` leave a bit to be desired.  It looks like I should rather do:

```smtp.sendmail(self.sender, self.to + self.cc + self.bcc, email_txt)```

And then remove the ```Bcc``` header as you suggested previously.

File: mailgun/models.py (hunk 2 above)
Comment (jimbobhickville) [inline mailgun/models.py]: Yeah, it's always a toss-up.  I like it because you can add fields to the model and they'll automatically just be included in the JSON response, which made it quicker to iterate to get the POC done.  But I can also see wanting to limit which fields are returned as some aren't relevant to the user (like ```topic``` is returned now but the user probably shouldn't be privy to that).

File: mailgun/processor.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    This is a blocking loop that doesn't exit until killed.
+    """
+    while True:
+        try:
+            # TODO: spin up a processor per topic, or iterate over topics?
+            # hardcoding to default topic for now due to time constraints
+            topic = const.DEFAULT_QUEUE
+            for message in queue.receive(topic):
+                msg_json = json.loads(message)
+                msg_obj = models.Messages.find(msg_json['message_id'])
```
Comment (russjones) [inline mailgun/processor.py]: It looks like you have everything you need to deliver a message in `message` that you pull out of Kafka. Why do you incur another database (`models.Messages.find` hit to to get the same data?

File: mailgun/api/messages.py (hunk 5 above)
Comment (russjones) [inline mailgun/api/messages.py]: That's what I was trying to get at, the `message_id` is really the only important field, why return the rest of the message (minus the body)?

File: mailgun/api/messages.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import falcon
+
+from mailgun import const
+from mailgun import exceptions as mg_exc
+from mailgun.models import Messages
+
+
+class MessageList(object):
+    """Controller for interacting with the full collection of messages"""
+    # TODO: input validation
```
Comment (russjones) [inline mailgun/api/messages.py]: What are you considering validating and how would you validate it?

File: mailgun/models.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+        :returns: list of `~mailgun.models.Message` objects
+        """
+        # TODO: should let them pass in the state and move it to the API
+        deleted_state = const.MessageState.DELETED
+        query = db_session.get_session().query(cls)
+        return query.filter(Messages.state != deleted_state).all()
+
+    @classmethod
+    def find_retryable(cls):
```
Comment (russjones) [inline mailgun/models.py]: It looks like you are using crate.io both as a message store and as a queue. Can you talk about the advantages and disadvantages of crate.io for both use cases?

File: mailgun/processor.py (hunk 6 above)
Comment (jimbobhickville) [inline mailgun/processor.py]: That was the original idea, but I needed to get the current ```state``` in case it was deleted, so I ended up loading it from the db.  Now that you bring it up, along with the other comment made about multiple processors, I think it would make more sense to rely on the built-in record versioning from crate to handle this: https://crate.io/docs/reference/sql/occ.html#sql-occ

So, I could just create the in-memory model from the data in Kafka, then issue the update to change state to ```PROCESSING``` by passing in the ```_version``` at the time the object was put on Kafka.  If the ```_version``` has since changed, just abort the processing and move along.  Whatever other update was made either resulted in a new message being put in to the queue and it will be processed again, or it means no processing is now required and we should leave it alone.  We'd need to add ```_version``` to the data put into Kafka to accomplish this, however, but it would handle both race conditions and avoiding an extra query here.  Would you like me to try to make the necessary changes for that?

File: mailgun/api/messages.py (hunk 7 above)
Comment (jimbobhickville) [inline mailgun/api/messages.py]: Validating the structure and content of the incoming JSON document.  In my current project we use the Python ```voluptuous``` library for that, but I've also seen ```jsonschema``` used in that manner.  There are undoubtedly many other options.  We made a decorator to make it easy to add the voluptuous-based validation (this is an example from my current project):

```
@json_validator(
        request,
        Any(
            action_data('node_groups', [{
                "count": All(int, Range(min=CONF.stack.component_min,
                                        max=CONF.stack.component_max)),
                "id": All(unicode, Length(min=1, max=36))
            }]),
            action_data('credentials', [{
                "name": unicode,
                "type": Any(SshKeysCred.cred_type,
                            S3Cred.cred_type,
                            CloudFilesCred.cred_type)
            }]),
            action_data('remove_credentials', [{
                "name": unicode,
                "type": SshKeysCred.cred_type
            }]),
            action_data('control_services', {
                'action': Any('start', 'stop', 'restart'),
                'services': [{'name': All(unicode)}],
            }),
        ),
        False,
        True)
```

I would prefer an approach that could be leveraged by documentation generation scripts to build out the expected document structure for the API docs.  I'd have to look into it more to see what's available there or build my own small layer over the top of an existing solution, if needed.

File: mailgun/models.py (hunk 8 above)
Comment (jimbobhickville) [inline mailgun/models.py]: I'm not really using it as a queue, though.  I guess in the sense that I'm setting that ```next_attempt``` time to force a requeuing in the future, sure, but it's not really a queue per se.  I view it solely as state persistence, just one piece of state data is when it should be re-processed.  I didn't want to rely on just sleeping a thread or something, as that would lose all the information about when to retry in the event of an outage, and I'm not aware of any existing message queues that let you tell it when to process a message.

File: mailgun/api/messages.py (hunk 5 above)
Comment (jimbobhickville) [inline mailgun/api/messages.py]: Partially habit born of having had to write API clients in the past.  I want a ```POST``` to return the same data structure that a ```GET``` would because the client can use the results of either verb with the same code paths (such as rendering the received data into a template) without the extra ```GET``` request to retrieve the remaining fields.  There are other auto-generated fields of import here, like ```state```, and potentially others to be added in the future. I guess I've just defaulted to this sort of pattern for so long that I hadn't really considered that it might be odd.  Is your concern that it's inefficient to just spit back at them what they just sent in?

File: mailgun/processor.py (hunk 6 above)
Comment (jimbobhickville) [inline mailgun/processor.py]: I explored this a bit.  SQLAlchemy even has a feature like this built in: http://docs.sqlalchemy.org/en/latest/orm/versioning.html#simple-version-counting

But it doesn't play nice with crate.io currently, nor is there any way to get at the underlying ```_version``` built-in.

Also, creating the model directly from the JSON has separate issues:

```
msg_obj = models.Message(**json_dict)
msg_obj.update(state=const.MessageState.PROCESSING)
```

Will fail because the session isn't aware that it's an existing record and attempts to do an ```INSERT``` rather than an ```UPDATE``` and hits a duplicate record error.  Trying to use ```session.merge(msg_obj, load=False)``` to add the in-memory object to the session fails with a different error unless you keep ```load``` set to True, in the which case you've now forced yourself to do a database query to verify that the in-memory object is legit. 

Basically, to do this cleanly I'd need to rip out sqlalchemy and start over with direct database interactions so I can take advantage of the ```_version``` field properly and avoid having to load it from the database before updating it.  With the amount of work required to do that, I'm not sure I can finish it in time.

**Comment by amatai** [comment]
Maybe I've missed this, how does data from Kafka/Crate get purged after the message has reached its final state? Please point me to it, if I missed it.

**Comment by jimbobhickville** [comment]
It doesn't get purged from Crate, it just gets the state updated to ```SENT``` or ```FAILED```.  I wasn't really sure what the retention policy should be so I opted to keep 'em around indefinitely for now.  Kafka auto-purges old messages based on a retention policy defined per topic, but it tracks the offset in zookeeper (automatically done via the ```KafkaConsumer``` object) to avoid re-processing the same message again.

File: mailgun/models.py (hunk 8 above)
Comment (russjones) [inline mailgun/models.py]: Okay, can you talk about the advantages and disadvantages of using crate.io for state persistence? Suppose we were going to use crate.io on our production servers, can you think of problems we might run into? Specifically can you discuss the design/architecture of crate.io and tell us why it may or may not be a good fit for state persistence?

File: mailgun/models.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+        If it succeeds, it sets the state to SENT.
+        If it fails, it either:
+            1. Sets the state to REJECTED so it gets retried
+            2. Sets the state to FAILED if all retries were exhausted.
+
+        The behavior for requeue time, number of retries, etc, is all
+        configurable.
+        """
+        if self.state not in const.PENDING_STATES:
```
Comment (amatai) [inline mailgun/models.py]: Under what condition do you think, this error can happen?

Comment (jimbobhickville) [inline mailgun/models.py]: The most likely condition that this check guards against is the message was set to ```DELETED``` by the consumer issuing a ```DELETE``` request to the API.  But it could also happen if we failed to record the current offset and re-processed a message from Kafka (such as if the processor crashed after updating the state in the crate database but before it recorded the offset).

File: mailgun/models.py (hunk 8 above)
Comment (jimbobhickville) [inline mailgun/models.py]: Crate's peer-to-peer architecture makes it easy to scale by automatically sharding data among the available nodes.  There is a peer-elected master process that controls re-replicating data when a node disappears, but if the master disappears, another is elected shortly.  This means a quorum of available nodes is required to maintain the cluster, otherwise queries will fail. I can't say that's any worse than traditional master/slave failover scenarios, but it's different and something to remember.  It also means you should always keep an odd number of nodes in a cluster so that there is no possibility of an evenly split scenario on a network partition.  While it's only eventually consistent for search indexes (i.e. ```SELECT``` *not* on the primary key), it does have immediate consistency on individual record ```GET``` calls by the primary key, so ```INSERT``` then ```SELECT``` or ```UPDATE``` then ```SELECT``` still work as expected, but it might now show up in a ```SELECT``` by any other field for "a few milliseconds" (according to their docs).  It offers optimistic concurrency control over row locking, which I think is sufficient for this use-case, but I haven't had enough experience with it to know for sure. It certainly beats having lock contention and/or starvation scenarios with traditional mutexes. The lack of transactions or relationships requires a different type of system design, but there wasn't really much need for either in this system so far.  The primary reason I chose it was because it supported both document storage and blob storage with automatic replication and efficient scaling, while still supporting SQL well enough to utilize sqlalchemy for the model definition.  That said, I realize it's still relatively new and unproven, and the sqlalchemy integration ended up being the wrong decision.

Let me know if you want me to go into any more depth on any specific item there; I think I covered most of the pros/cons.

File: mailgun/models.py (hunk 4 above)
Comment (anton-efimenko) [inline mailgun/models.py]: Yeah, sorry, my bad.

File: mailgun/models.py (hunk 8 above)
Comment (anton-efimenko) [inline mailgun/models.py]: That's really interesting. Could you elaborate on sqlalchemy integration being the wrong decision? What particularly do you dislike in the way it turned out?

File: mailgun/models.py (hunk 4 above)
Comment (anton-efimenko) [inline mailgun/models.py]: Regarding running multiple processes. If 2 processes can get to this point and both grab the same message, one of them would be rejected on the re-queuing. Is it a correct assumption?

File: mailgun/models.py (hunk 8 above)
Comment (jimbobhickville) [inline mailgun/models.py]: The conversation with russjones here elaborates a bit: https://github.com/jimbobhickville/mini-mailgun/pull/1/files#diff-d9b03ffbcee7d3336eb8841cde2b4cecR42

There were 2 main issues:

1.The concurrency controls in crate are not integrated well with sqlalchemy at this point. This could be fixed with some patches to sqlalchemy-crate to fix it to utilize crate's "optimistic concurrency controls" when you specify a ```version_id_col``` (a similar feature of sqlalchemy).
2. The way sqlalchemy tracks objects in a session makes it not work well for doing an ```UPDATE``` without first doing a ```SELECT```.  This negates the optimization I made for putting all of the message details into the kafka queue directly, in the case the message is small enough to fit, because it has to do an additional query to load the details from crate again before it can update them.

I *could* just use the sql generation bits from sqlalchemy, but if I'm not going to take advantage of the session handling and ORM layers, I might as well just use the crate client directly.  The underlying queries I'd need to write aren't that complex.

Comment (russjones) [inline mailgun/models.py]: Well let me be more specific, can go into more details about how data is stored in crate.io? Specifically, I was looking at the section on [storage and consistency](https://crate.io/docs/reference/storage_consistency.html), and saw this:

```
Lucene only appends data to segment files, which means that data written to the disc will
never be mutated. This makes it easy for replication and recovery, since syncing a shard
is simply a matter of fetching data from a specific marker.
```

Can you talk about the advantages/disadvantages of using a database like this in the manner that you do (lots of updates and deletes).

Comment (jimbobhickville) [inline mailgun/models.py]: Ah, I see.  Honestly, I missed that tidbit when I was researching crate. There is no mention in the docs if it ever compacts old, outdated data, but it sounds like not since it would only be able to do so if an entire segment file was able to be reaped.  So there's definitely a scaling issue as updates involve only adding data and never removing it.  It doesn't mention if the whole record is recreated on an update or only the diff.  If it's just the diff, there's additional overhead on reads as it has to find all of the diffs and apply them to produce the current version of the record.  If it's the whole record, reads will be faster, but it will use more disk space to store all the copies, meaning we have to make sure we keep a better eye on disk usage and scale appropriately.

One thing I could do is move the static data into a separate table so that updates are only updating a record that has a smaller footprint.  That would help limit how much disk space was utilized.  Another option might be to pre-generate the email message and store that in the blob store, so then the ```subject``` and ```body``` fields would not even be in crate at all.  The happy path involves 1 insert and 2 updates, and worst-case is 1 insert and 8 updates (2 updates per attempt * 3 attempts + 1 update per requeue * 2 requeues), so that might work out ok if each update is only a small amount of data.  There are no deletes in the codebase (the HTTP ```DELETE``` simply updates the ```state``` to ```DELETED``` and the ```GET``` only returns messages that aren't in that state).

Further down the page, it mentions that writes actually are synchronous to all replicas.  I had also glossed over that fact because elsewhere it mentioned eventual consistency.  This means that write performance is impacted by the number of replicas we have set, so we'd need to be considerate of that.  More copies == more safety, but slower write performance. I'd have to do some benchmarking to see if it could even keep up given the load we expect on the production system.

File: mailgun/models.py (hunk 4 above)
Comment (jimbobhickville) [inline mailgun/models.py]: Sorry I overlooked this message yesterday.  So this one wouldn't work with multiple processes as-is.  If I got the versioning bits working, then yes, it would reject it when I tried to update the state back to ```PENDING```, but all of the processes would be fighting over the same records.  I'd have to do some sort of chunking of the data, maybe spin up each process to grab some range of UUIDs (like 'starts with A, starts with B', etc).  I don't know what rejection rates are like to know how well this part would actually need to scale.  Since it auto-recovers on resumption, we could just limit it to a single process until we knew it was unable to keep up and then work out a better scaling strategy going forward.

File: mailgun/models.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+        :param message_id: the uuid of the message to be found
+        :type message_id: str
+        :returns: `~mailgun.models.Message` object
+        """
+        query = db_session.get_session().query(cls)
+        return query.filter_by(message_id=message_id).one()
+
+    @classmethod
+    def find_all(cls):
```
Comment (horkhe) [inline mailgun/models.py]: Do you see any issues with this method?

Comment (jimbobhickville) [inline mailgun/models.py]: Yes, it isn't very memory-efficient because it slurps the whole result set into memory.  Because I was just doing that to generate the ```JSON``` response, I didn't bother to optimize it here.  There are two things I *should* do:

1. Convert this to a generator that yields individual ```Message``` objects so you don't unnecessarily load all of them into memory at once.
2. Allow the user to limit the size of the result set with pagination parameters. Note: this limiting of the resultset should be done on the database server, not in this client code.  I say this because some codebases I've seen have done pagination in the client.
